src/jit/optimizer.cpp

```diff
@@ -6830,6 +6830,30 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
         {
             GenTree* tree = *use;
 
+            if (tree->OperIsLocal())
+            {
+                GenTreeLclVarCommon* lclVar = tree->AsLclVarCommon();
+                unsigned             lclNum = lclVar->GetLclNum();
+
+                // To be invariant a LclVar node must not be the LHS of an assignment ...
+                bool isInvariant = !user->OperIs(GT_ASG) || (user->AsOp()->gtGetOp1() != tree);
+                // and the variable must be in SSA ...
+                isInvariant = isInvariant && m_compiler->lvaInSsa(lclNum);
+                // and the SSA definition must be outside the loop we're hoisting from.
+                isInvariant = isInvariant &&
+                              !m_compiler->optLoopTable[m_loopNum].lpContains(
+                                  m_compiler->lvaGetDesc(lclNum)->GetPerSsaData(lclVar->GetSsaNum())->m_defLoc.m_blk);
+
+                if (isInvariant)
+                {
+                    Value& top = m_valueStack.TopRef();
+                    assert(top.Node() == tree);
+                    top.m_invariant = true;
```

Comment:
I've re-ran diffs since it's been a while and found a small issue that I've missed before, apparently it only affects x86. Because for locals all the logic below is skipped, this should set `m_hoistable` to `optHoistCandidate(tree)` here, to handle those (rather rare) cases where a local node is worth hoisting (typically LCL_FLDs because they cannot be enregistered).

Reply:
Do we need to Pop/check children here as well?

Reply:
No because local nodes have no children.

Reply:
To maintain 0 diffs on x86 this also needs to check `optVNIsLoopInvariant`, even though doing that is pretty much bogus. There is one diff in the x86 corelib that's caused by the new code being less conservative (but still correct) than the old code:
* a `char` method argument gets "reinterpreted" via a `LCL_FLD` (not sure why, that's some x86 oddity that's not related to loop hoisting)
* this `LCL_FLD` does not have a field sequence so it gets a "new, unique VN" which is associated with the block where the `LCL_FLD` is, despite the fact that the definition of the method argument is considered to be at the beginning of the first block in the method.
* the block where the `LCL_FLD` occurs is inside the loop so `optVNIsLoopInvariant` thinks that the "new, unique VN" isn't invariant and this incorrectly blocks `LCL_FLD` hoisting.

IMO this is a mess and it should be cleaned up at some point. But right now I'd prefer to keep the diffs to a minimum - only those diffs that arise due to changes in hoisting order.

Reply:
Are you going to do this change in this PR?

Reply:
Yes, I've just pushed a commit that brings this to 0 diffs on x86. Checking ARM diffs now.

Reply:
> IMO this is a mess and it should be cleaned up at some point

Are you referring to this strange situation around the `char` method argument that's reinterpreted via `LCL_FLD`, or something broader than that? (Mostly just curious, but it would be good to know, especially if you have some insight as to what's causing it).

Reply:
That `LCL_FLD` case was just an example I encountered, the fix I pushed simply restores the original invariance logic for local nodes - the local node is invariant if both SSA and VN say so. 

The SSA part is probably clear to everyone. The VN part is is rather dubious - what does VN know about loops? 

Normally not a lot but #6747 associated basic blocks to VNs such that we can tell where in the flow graph a "unique"/"opaque" VN occurred. It turns out that that change doesn't handle all cases - sometimes the VN associated block should be the SSA definition block, not the SSA use block.

But the bigger question is why is VN even involved in loop hoisting. If you have `a + b` and SSA says that `a` and `b` are invariant then you don't need VN to tell you that `a + b` is also invariant. Best I can tell, the main reason VN is involved is that it can tell you if an indirection is invariant. Since loop hoisting was written, a rudimentary memory SSA form was added but that's unlikely to be enough to replace VN in loop hoisting as VN also acts as some kind of alias analysis.